### PR TITLE
Handle exponents in FixedDecimal::from_str()

### PR DESCRIPTION
**Pull request by Manishearth**
Needed for https://github.com/unicode-org/icu4x/issues/166

### CODE REVIEW

File: utils/fixed_decimal/src/decimal.rs [hunk 1]
```diff
@@ -570,7 +598,7 @@ impl FromStr for FixedDecimal {
 
         // Constructing DecimalFixed.digits
         let mut v: SmallVec<[u8; 8]> = SmallVec::with_capacity(digits_str_len);
-        for c in no_sign_str[leftmost_digit..rightmost_digit].iter() {
+        for c in no_exponent_str[leftmost_digit..rightmost_digit].iter() {
```
Comment (zbraniecki) [inline utils/fixed_decimal/src/decimal.rs:601]: `question`: `filter_map`?

File: utils/fixed_decimal/src/decimal.rs [hunk 2]
```diff
@@ -580,6 +608,31 @@ impl FromStr for FixedDecimal {
         debug_assert_eq!(v_len, digits_str_len);
         dec.digits = v;
 
+        // Extract the exponent part
+        if has_exponent {
+            let mut pow = 0;
+            let mut pos_neg = 1;
+            for digit in &no_sign_str[exponent_index + 1..] {
```
Comment (zbraniecki) [inline utils/fixed_decimal/src/decimal.rs:615]: `question`: `reduce`?

**Review by zbraniecki — APPROVED** [review]
logic looks good and the code is readable. I think Rust offers a number of nice APIs for iterating that would make it even cleaner, but nothing blocking, so up to Manish to include or reject my suggestions.

File: utils/fixed_decimal/src/decimal.rs [hunk 3]
```diff
@@ -461,42 +461,70 @@ impl FromStr for FixedDecimal {
...
+        // dot_index: gives the index of dot (after removing the sign) -- equal to length of
+        // the no_sign_str if there is no dot
+        // exponent_index: gives the index of exponent (after removing the sign) -- equal to length of
         // the no_sign_str if there is no dot
-        let no_sign_str_len = no_sign_str.len();
         let mut has_dot = false;
-        let mut dot_index = no_sign_str_len;
+        let mut has_exponent = false;
+        let mut dot_index = no_sign_str.len();
+        let mut exponent_index = no_sign_str.len();
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply utils/fixed_decimal/src/decimal.rs:472]: We still use `exponent_index` even if it's an option, though. I can rename that to `exponent_or_end_index`

File: utils/fixed_decimal/src/decimal.rs [hunk 4]
```diff
@@ -461,42 +461,70 @@ impl FromStr for FixedDecimal {
...
+                    return Err(Error::Syntax);
+                }
+            } else if *c == b'-' {
+                // Allow a single minus sign after the exponent
+                if has_exponent && exponent_index == i - 1 {
+                    continue;
+                } else {
+                    return Err(Error::Syntax);
                 }
             } else if *c < b'0' || *c > b'9' {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply utils/fixed_decimal/src/decimal.rs:505]: We don't have a `char`

[the remark this answers is not in the document]
Comment (zbraniecki) [reply utils/fixed_decimal/src/decimal.rs:505]: https://doc.rust-lang.org/std/primitive.u8.html#method.is_ascii_digit

File: utils/fixed_decimal/src/decimal.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zbraniecki) [reply utils/fixed_decimal/src/decimal.rs:472]: Isn't `expontent.is_none()` equivalent of `no_sign_str.len()` then?

**Review by zbraniecki — DISMISSED** [review]
lgtm! (I still like the `Option(index)` more, but leave it to the PR author discretion).

File: utils/fixed_decimal/src/decimal.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply utils/fixed_decimal/src/decimal.rs:472]: Yes, I mean that every time we use `exponent_index` we want to use `exponent.unwrap_or(no_sign_str.len())` anyway, so it's not worth it to store the Option

**Comment by Manishearth** [comment]
Yeah overall that code is existing code and I feel like it's just as complex with the Option, with the complexities shifting around

**Review by zbraniecki — APPROVED** [review]

**Comment by sffc** [comment]
I want to discuss scientific notation from an architecture perspective. Filed #1267.

**Pull request merged**